### PR TITLE
Implement Set builtin object

### LINKED ISSUE #400: Implement Set()
**ECMASCript feature**
Set objects are collections of ECMAScript language values. A distinct value may only occur once as an element of a Set's collection. Distinct values are discriminated using the SameValueZero comparison algorithm.

I would like to see `Set` implemented. [ECMAScript specification][spec].

[spec]: https://tc39.es/ecma262/#sec-set-objects

**Example code**
```javascript
let mySet = new Set()

mySet.add(1)           // Set [ 1 ]
mySet.add(5)           // Set [ 1, 5 ]
mySet.add(5)           // Set [ 1, 5 ]
mySet.add('some text') // Set [ 1, 5, 'some text' ]
let o = {a: 1, b: 2}
mySet.add(o)

mySet.add({a: 1, b: 2})   // o is referencing a different object, so this is okay

mySet.has(1)              // true
mySet.has(3)              // false, since 3 has not been added to the set
mySet.has(5)              // true
mySet.has(Math.sqrt(25))  // true
mySet.has('Some Text'.toLowerCase()) // true
mySet.has(o)       // true

mySet.size         // 5

mySet.delete(5)    // removes 5 from the set
mySet.has(5)       // false, 5 has been removed

mySet.size         // 4, since we just removed one value

console.log(mySet)
// logs Set(4) [ 1, "some text", {…}, {…} ] in Firefox
// logs Set(4) { 1, "some text", {…}, {…} } in Chrome
```

**Example to work from**
Array is implemented here: https://github.com/jasonwilliams/boa/blob/master/boa/src/builtins/array/mod.rs

**Contributing**
https://github.com/jasonwilliams/boa/blob/master/CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by RageKnify**
This Pull Request fixes/closes #400.

Besides **implementing `Set`** this PR also changes the following:

- Change `Value::hash` so that `1i32 == 1f64`
- Fix `Map.prototype.length` (was 1, should be 0 because the argument is optional)
- Fix msg for invoking Map without new (copied V8's message, gives more information)
- Simplify calculation of map size when displaying Maps (get size from the `OrderedMap` rather than going throught the property
- Introduce IteratorClose
- Fix switching between PropertyDescriptor kinds
- Fix deletion of accessor properties
- Add equality check for `Value::Symbol`

The change to `Value::hash` was a bandaid fix to the fact that incrementing/decrementing integers with `++`/`--` results in a float, that should still be addressed, but I think this change would still be needed.

ps: I'll leave the remaining tests failing, currently failing tests [19] can be grouped by:
- tests that expect `WeakSet` to be implemented [14]
- tests that expect correct implementation of `Set.prototype.forEach` when elements are removed (tracked in #1092 for `Map`, solution will likely be easily translated to `Set`) [2]
- test that expects `this` to be `undefined` in strict mode [1]
- tests that expect `$262` to be defined and use `$262.createRealm()` [2]
- ~~tests that expect `Reflect` to be implemented, which should pass once #1033 is merged [2]~~ (tested locally, they pass now that #1033 has been merged)

### CODE REVIEW

File: boa/src/builtins/set/set_iterator.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-set-iterator-objects
+#[derive(Debug, Clone, Finalize, Trace)]
+pub struct SetIterator {
+    iterated_set: Value,
+    set_next_index: usize,
+    set_iteration_kind: SetIterationKind,
```
Comment (HalidOdat) [inline boa/src/builtins/set/set_iterator.rs:28]:
```suggestion
    next_index: usize,
    iteration_kind: SetIterationKind,
```
I think it's clearer if we don't prefix the fields with `set_` since they are already in the `SetInterator`.

File: boa/src/builtins/set/set_iterator.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        set: Value,
+        kind: SetIterationKind,
+    ) -> Result<Value> {
+        let set_iterator = Value::new_object(context);
+        set_iterator.set_data(ObjectData::SetIterator(Self::new(set, kind)));
+        set_iterator
+            .as_object()
+            .expect("set iterator object")
+            .set_prototype_instance(context.iterator_prototypes().set_iterator().into());
+        Ok(set_iterator)
```
Comment (HalidOdat) [inline boa/src/builtins/set/set_iterator.rs:61]:
```suggestion
    ) -> Value {
        let set_iterator = Value::new_object(context);
        set_iterator.set_data(ObjectData::SetIterator(Self::new(set, kind)));
        set_iterator
            .as_object()
            .expect("set iterator object")
            .set_prototype_instance(context.iterator_prototypes().set_iterator().into());
        set_iterator
```
Since we cant throw I think its best to remove the unnecessary Result wrap.

File: boa/src/builtins/map/mod.rs [hunk 3]
```diff
@@ -75,7 +74,7 @@ impl Map {
         context: &mut Context,
     ) -> Result<Value> {
         if new_target.is_undefined() {
-            return context.throw_type_error("Map requires new");
+            return context.throw_type_error("Constructor Map requires 'new'");
```
Comment (Razican) [inline boa/src/builtins/map/mod.rs:77]: I would rephrase this a bit:

```suggestion
            return context.throw_type_error("calling the Map requires using 'new'");
```

An example of Firefox, that seems a bit more clarifying:
```
calling a builtin Map constructor without new is forbidden
```

This also requires changing the test.

File: boa/src/builtins/set/ordered_set.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    V: Hash + Eq;
+
+impl<V: Eq + Hash + Trace, S: BuildHasher> Finalize for OrderedSet<V, S> {}
+unsafe impl<V: Eq + Hash + Trace, S: BuildHasher> Trace for OrderedSet<V, S> {
+    custom_trace!(this, {
+        for v in this.0.iter() {
+            mark(v);
+        }
+    });
+}
```
Comment (Razican) [inline boa/src/builtins/set/ordered_set.rs:25]: Since this is implementing a custom trace, I would add a comment on why this is needed and how it works.

**Review by Razican — APPROVED** [review]
Looks very good!! I added some suggestions, but it's good to go from my side :) congrats!

File: boa/src/builtins/map/mod.rs (hunk 3 above)
Comment (RageKnify) [reply boa/src/builtins/map/mod.rs:77]: I'll be changing to Firefox's style, the one I had set was Chromium's/V8's, but I agree FF's gives more context.

File: boa/src/builtins/set/ordered_set.rs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    }
+}
+
+impl<V> OrderedSet<V>
+where
+    V: Hash + Eq,
+{
+    pub fn new() -> Self {
+        OrderedSet(IndexSet::new())
+    }
```
[the remark this answers is not in the document]
Comment (RageKnify) [reply boa/src/builtins/set/ordered_set.rs:45]: Do you mean `OrderedSet(IndexSet::default())`? We could but I don't think we gain anything with it...

File: boa/src/builtins/set/mod.rs [hunk 6]
```diff
@@ -0,0 +1,407 @@
+use crate::{
```
Comment (HalidOdat) [inline boa/src/builtins/set/mod.rs:1]: Maybe add some documentation like other builtin global objects.

**Review by HalidOdat — APPROVED** [review]
Looks good to me! Jest some minor documentation things.

File: boa/src/builtins/set/ordered_set.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Razican) [reply boa/src/builtins/set/ordered_set.rs:45]: Since it's a private structure, I think it's good as-is

File: boa/src/builtins/set/set_iterator.rs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+impl SetIterator {
+    pub(crate) const NAME: &'static str = "SetIterator";
+
+    fn new(set: Value, kind: SetIterationKind) -> Self {
+        SetIterator {
+            iterated_set: set,
+            next_index: 0,
+            iteration_kind: kind,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Razican) [reply boa/src/builtins/set/set_iterator.rs:40]: Just something on the lines:
```
This creates a new `SetIterator`, that will start on position with index 0.
```

File: boa/src/builtins/set/ordered_set.rs (hunk 4 above)
Comment (Razican) [reply boa/src/builtins/set/ordered_set.rs:25]: Hmmm maybe @HalidOdat knows about this more than me.

**Comment by Razican** [comment]
The build failure is due to changes in `prettier`. A re-base should fix it (if we want, if not, I'm happy to merge it with this failure)

File: boa/src/builtins/set/ordered_set.rs (hunk 4 above)
Comment (HalidOdat) [reply boa/src/builtins/set/ordered_set.rs:25]: > Hmmm maybe @HalidOdat knows about this more than me.

It is needed because `IndexSet` is does not implement `Trace`, that's why we wrap it in a struct that does implement it (`OrderedSet`). The custom trace iterates through all the values in the set and calls trace on them (the `mark(v)` part).

File: boa/src/builtins/set/mod.rs [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+                .throw_type_error("calling a builtin Set constructor without new is forbidden");
+        }
+
+        // 2
+        let set_prototype = context
+            .global_object()
+            .get(&"Set".into(), context.global_object().into(), context)?
+            .get_field(PROTOTYPE, context)?
+            .as_object()
+            .expect("'Set' global property should be an object");
```
Comment (HalidOdat) [inline boa/src/builtins/set/mod.rs:121]: we should store the constructor and prototype in the [`StandardOobjects`](https://github.com/boa-dev/boa/blob/master/boa/src/context.rs#L80-L99).

Comment (HalidOdat) [reply boa/src/builtins/set/mod.rs:121]: then we only need to do:
```rust
let prototype = context.standard_objects().set_object().prototype();
```

File: boa/src/builtins/set/mod.rs [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+        let to_string_tag = WellKnownSymbols::to_string_tag();
+
+        let values_function = FunctionBuilder::new(context, Self::values)
+            .name("values")
+            .length(0)
+            .callable(true)
+            .constructable(false)
+            .build();
+
+        let set_object = ConstructorBuilder::new(context, Self::constructor)
```
Comment (HalidOdat) [inline boa/src/builtins/set/mod.rs:67]:
```suggestion
        let set_object = ConstructorBuilder::with_standard_object(
            context,
            Self::constructor,
            context.standard_objects().set_object().clone(),
        )
```
Here we should init the global Set object

File: boa/src/builtins/set/mod.rs (hunk 8 above)
Comment (RageKnify) [reply boa/src/builtins/set/mod.rs:121]: Can you explain a bit more why we want this? Is this so that people using boa can get these StandardObjects quickly from the Context instead of doing a lookup in a HashMap?

**Comment by RageKnify** [comment]
The MacOS CI "failure" was just a problem on GHs side, tests have been working so far, no reason to expect them to fail now.

**Review by HalidOdat — APPROVED** [review]
